File: projects/comprehensive_metal_ecology/scripts/subcategory_lambda_meta.R
#!/usr/bin/env Rscript
# Random-effects meta-regression of Pagel's lambda across KEGG functional subcategories
# Reference: Hadfield & Nakagawa 2010 J Evol Biol 23:494-508
#
# Model: lambda_i ~ subcategory (fixed moderator) + u_i (random study effect)
# Sampling variance: vi = 1 / (n_genera - 3)  [approximation treating lambda like
#   a partial correlation; exact sampling variance requires bootstrap from caper::pgls]
# Method: REML (restricted maximum likelihood)
#
# Excludes subcategory == "Unknown" (uninformative moderator level; 121 KOs)

suppressMessages(library(metafor))

DATA  <- "projects/comprehensive_metal_ecology/data"
FIGS  <- "projects/comprehensive_metal_ecology/figures"

# ── Load data ──────────────────────────────────────────────────────────────────
df <- read.csv(file.path(DATA, "phylo_d_all_ko.csv"), stringsAsFactors = FALSE)
cat(sprintf("Loaded %d metal KOs\n", nrow(df)))
cat(sprintf("Subcategory distribution:\n"))
print(table(df$subcategory))

# ── Exclude Unknown subcategory ────────────────────────────────────────────────
df_known <- df[df$subcategory != "Unknown", ]
cat(sprintf("\nAfter excluding Unknown: %d KOs\n", nrow(df_known)))

# ── Sampling variance approximation ───────────────────────────────────────────
# vi = 1 / (n - 3): analogous to Fisher-Z sampling variance for correlations.
# Lambda ~ Pearson r^2 of trait evolution, so this is conservative (slightly
# inflated vi for most KOs).
df_known$vi <- 1 / pmax(df_known$n_genera - 3, 1)

# ── Relevel: Cofactor Biosynthesis as reference (biological anchor) ──────────
df_known$subcategory <- factor(
  df_known$subcategory,
  levels = c("Cofactor Biosynthesis", "Metal-dependent Metabolism",
             "Resistance/Detoxification", "Sensing/Regulation",
             "Transport/Homeostasis")
)

# ── Model 1: Intercept-only random effects (overall mean lambda) ──────────────
cat("\n=== Model 1: Overall mean lambda (intercept-only) ===\n")
m0 <- rma(yi = lambda, vi = vi, data = df_known, method = "REML")
print(m0)

# ── Model 2: Subcategory as moderator ─────────────────────────────────────────
cat("\n=== Model 2: Subcategory moderator ===\n")
m1 <- rma(yi = lambda, vi = vi, mods = ~ subcategory, data = df_known, method = "REML")
print(m1)

# ── Likelihood ratio test: does subcategory improve fit? ──────────────────────
cat("\n=== LRT: M1 vs M0 ===\n")
lrt <- anova(m1, m0)
print(lrt)

# ── Per-subcategory predicted means with 95% CI ───────────────────────────────
cat("\n=== Per-subcategory predicted lambda (from M2) ===\n")
cats   <- levels(df_known$subcategory)
X_pred <- model.matrix(~ subcategory,
                        data = data.frame(subcategory = factor(cats, levels = cats)))
preds  <- predict(m1, newmods = X_pred[, -1, drop = FALSE])  # drop intercept col
pred_df <- data.frame(
  subcategory = cats,
  pred_lambda = round(preds$pred, 4),
  ci_lo       = round(preds$ci.lb, 4),
  ci_hi       = round(preds$ci.ub, 4),
  n_ko        = as.integer(table(df_known$subcategory))
)
print(pred_df)

# ── Save results ───────────────────────────────────────────────────────────────
out <- data.frame(
  model          = c("intercept_only", "subcategory_moderator"),
  QM             = c(NA, round(m1$QM, 3)),
  QM_df          = c(NA, m1$p - 1L),
  QMp            = c(NA, round(m1$QMp, 4)),
  tau2           = c(round(m0$tau2, 4), round(m1$tau2, 4)),
  I2             = c(round(m0$I2, 1), round(m1$I2, 1)),
  R2             = c(NA, round(m1$R2, 3)),
  LRT_Q          = c(NA, round(lrt$LRT, 3)),
  LRT_df         = c(NA, lrt$p.diff),
  LRT_p          = c(NA, round(lrt$pval, 4)),
  n_KO           = c(nrow(df_known), nrow(df_known)),
  n_KO_unknown   = c(sum(df$subcategory == "Unknown"), sum(df$subcategory == "Unknown"))
)
write.csv(out, file.path(DATA, "subcategory_meta_analysis.csv"), row.names = FALSE)

write.csv(pred_df, file.path(DATA, "subcategory_predicted_lambda.csv"), row.names = FALSE)

cat(sprintf("\nSaved -> %s/subcategory_meta_analysis.csv\n", DATA))
cat(sprintf("Saved -> %s/subcategory_predicted_lambda.csv\n", DATA))

# ── Summary for REPORT ────────────────────────────────────────────────────────
cat("\n=== REPORT SUMMARY ===\n")
cat(sprintf("Overall mean lambda (intercept-only): %.3f [%.3f, %.3f]\n",
            m0$beta[1], m0$ci.lb, m0$ci.ub))
cat(sprintf("tau2 = %.4f, I2 = %.1f%%\n", m0$tau2, m0$I2))
cat(sprintf("\nSubcategory moderator test: QM(%d) = %.2f, p = %.4f\n",
            m1$p - 1L, m1$QM, m1$QMp))
cat(sprintf("R2 (variance explained by subcategory): %.1f%%\n", m1$R2))
cat(sprintf("\nOriginal KW test: H=8.71, p=0.1212 (unweighted, ignored sampling variances)\n"))

File: projects/comprehensive_metal_ecology/scripts/subcategory_lambda_meta.py
"""
Random-effects meta-regression of Pagel's lambda across KEGG functional subcategories.

Reference: Hadfield & Nakagawa 2010 J Evol Biol 23:494-508;
           DerSimonian & Laird 1986 Control Clin Trials 7:177-188

Model: lambda_i ~ subcategory (fixed moderator) + u_i (random KO effect)
Sampling variance: vi = 1 / (n_genera - 3)
tau2: DerSimonian-Laird method-of-moments estimate
"""

import numpy as np
import pandas as pd
from scipy import stats
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
DATA = ROOT / "projects/comprehensive_metal_ecology/data"

# ── Load data ──────────────────────────────────────────────────────────────────
df = pd.read_csv(DATA / "phylo_d_all_ko.csv")
print(f"Loaded {len(df)} metal KOs")
print(df['subcategory'].value_counts())

df = df[df['subcategory'] != 'Unknown'].copy()
print(f"\nAfter excluding Unknown: {len(df)} KOs")

# ── Sampling variance approximation ───────────────────────────────────────────
df['vi'] = 1.0 / (df['n_genera'] - 3).clip(lower=1)

# ── Reference level and design matrix ─────────────────────────────────────────
cats_ordered = ['Cofactor Biosynthesis', 'Metal-dependent Metabolism',
                'Resistance/Detoxification', 'Sensing/Regulation',
                'Transport/Homeostasis']
df['subcategory'] = pd.Categorical(df['subcategory'], categories=cats_ordered)

yi = df['lambda'].values
vi = df['vi'].values

# Design matrix (subcategory dummy, Cofactor Biosynthesis = intercept)
X = np.column_stack([
    np.ones(len(df)),
    (df['subcategory'] == 'Metal-dependent Metabolism').astype(float).values,
    (df['subcategory'] == 'Resistance/Detoxification').astype(float).values,
    (df['subcategory'] == 'Sensing/Regulation').astype(float).values,
    (df['subcategory'] == 'Transport/Homeostasis').astype(float).values,
])
col_names = ['(Intercept)', 'Metal-dep_Metab', 'Resistance/Detox',
             'Sensing/Reg', 'Transport/Homeo']

def fit_wls(yi, vi_total, X):
    """Weighted least squares with weights 1/vi_total."""
    W = np.diag(1.0 / vi_total)
    XtW = X.T @ W
    XtWX_inv = np.linalg.inv(XtW @ X)
    beta = XtWX_inv @ (XtW @ yi)
    vcov = XtWX_inv
    return beta, vcov

def dl_tau2(yi, vi, X):
    """DerSimonian-Laird method-of-moments estimate of tau2."""
    W0 = np.diag(1.0 / vi)
    beta0, _ = fit_wls(yi, vi, X)
    resid = yi - X @ beta0
    Q = resid @ W0 @ resid  # Cochran Q
    p = X.shape[1]
    k = len(yi)
    # Trace term for DL formula
    XtW0X = X.T @ W0 @ X
    c = np.trace(W0) - np.trace(np.linalg.inv(XtW0X) @ X.T @ W0 @ W0 @ X)
    tau2 = max(0.0, (Q - (k - p)) / c)
    return tau2, Q, k - p

# ── Model 0: Intercept only ────────────────────────────────────────────────────
X0 = np.ones((len(df), 1))
tau2_0, Q0, df0 = dl_tau2(yi, vi, X0)
beta0, vcov0 = fit_wls(yi, vi + tau2_0, X0)
se0 = np.sqrt(np.diag(vcov0))
print(f"\n=== Model 0: Intercept-only ===")
print(f"Overall mean lambda = {beta0[0]:.4f} [{beta0[0]-1.96*se0[0]:.4f}, {beta0[0]+1.96*se0[0]:.4f}]")
print(f"tau2 = {tau2_0:.4f}")
# I2 = tau2 / (tau2 + typical_vi)
typical_vi = np.median(vi)
I2_0 = 100 * tau2_0 / (tau2_0 + typical_vi)
print(f"I2 (approx) = {I2_0:.1f}%")
print(f"Cochran Q({df0}) = {Q0:.2f}")

# ── Model 1: Subcategory moderator ────────────────────────────────────────────
tau2_1, Q1_residual, df1_resid = dl_tau2(yi, vi, X)
beta1, vcov1 = fit_wls(yi, vi + tau2_1, X)
se1 = np.sqrt(np.diag(vcov1))
z1 = beta1 / se1
p1 = 2 * (1 - stats.norm.cdf(np.abs(z1)))

print(f"\n=== Model 1: Subcategory moderator ===")
print(f"tau2 = {tau2_1:.4f}")
I2_1 = 100 * tau2_1 / (tau2_1 + typical_vi)
print(f"I2 (approx) = {I2_1:.1f}%")
print(f"Residual Q({df1_resid}) = {Q1_residual:.2f}")

print(f"\nCoefficients:")
for name, b, se, z, p in zip(col_names, beta1, se1, z1, p1):
    stars = '***' if p < 0.001 else '**' if p < 0.01 else '*' if p < 0.05 else 'ns'
    print(f"  {name:<30s}  b={b:+.4f}  SE={se:.4f}  z={z:+.2f}  p={p:.4f}  {stars}")

# ── Omnibus moderator test (Wald QM) ──────────────────────────────────────────
# Test the 4 slope coefficients jointly
W_star = np.diag(1.0 / (vi + tau2_1))
XtWsX = X.T @ W_star @ X
beta_slopes = beta1[1:]
# Submatrix for slopes
vcov_slopes = np.linalg.inv(XtWsX[1:, 1:])
QM = beta_slopes @ np.linalg.solve(vcov_slopes, beta_slopes)
df_QM = len(beta_slopes)
p_QM = 1 - stats.chi2.cdf(QM, df_QM)
print(f"\n=== Omnibus moderator test (Wald) ===")
print(f"QM({df_QM}) = {QM:.3f}, p = {p_QM:.4f}")

# ── R2 analog (variance explained by subcategory) ─────────────────────────────
R2 = max(0.0, (tau2_0 - tau2_1) / tau2_0) if tau2_0 > 0 else 0.0
print(f"\nR2 analog (variance explained by subcategory): {R2*100:.1f}%")

# ── Per-subcategory predicted means ───────────────────────────────────────────
print(f"\n=== Per-subcategory predicted lambda ===")
pred_rows = []
for i, cat in enumerate(cats_ordered):
    x_row = np.zeros(len(beta1))
    x_row[0] = 1.0
    if i > 0:
        x_row[i] = 1.0
    pred_mean = float(x_row @ beta1)
    pred_se   = np.sqrt(x_row @ vcov1 @ x_row)
    n_ko      = (df['subcategory'] == cat).sum()
    pred_rows.append({
        'subcategory': cat,
        'pred_lambda': round(pred_mean, 4),
        'ci_lo': round(pred_mean - 1.96 * pred_se, 4),
        'ci_hi': round(pred_mean + 1.96 * pred_se, 4),
        'n_ko': int(n_ko)
    })
    print(f"  {cat:<35s}: {pred_mean:.4f} [{pred_mean-1.96*pred_se:.4f}, {pred_mean+1.96*pred_se:.4f}]  n={n_ko}")

pred_df = pd.DataFrame(pred_rows)

# ── Save ───────────────────────────────────────────────────────────────────────
summary_df = pd.DataFrame([{
    'model': 'intercept_only',
    'QM': None, 'QM_df': None, 'QMp': None,
    'tau2': round(tau2_0, 4), 'I2': round(I2_0, 1),
    'R2': None, 'n_KO': len(df), 'n_KO_unknown': (df['subcategory'].isna()).sum()
}, {
    'model': 'subcategory_moderator',
    'QM': round(QM, 3), 'QM_df': df_QM, 'QMp': round(p_QM, 4),
    'tau2': round(tau2_1, 4), 'I2': round(I2_1, 1),
    'R2': round(R2, 3), 'n_KO': len(df), 'n_KO_unknown': 121
}])
summary_df.to_csv(DATA / "subcategory_meta_analysis.csv", index=False)
pred_df.to_csv(DATA / "subcategory_predicted_lambda.csv", index=False)
print(f"\nSaved -> {DATA}/subcategory_meta_analysis.csv")
print(f"Saved -> {DATA}/subcategory_predicted_lambda.csv")

print(f"\n=== REPORT SUMMARY ===")
print(f"Overall mean lambda: {beta0[0]:.3f} [{beta0[0]-1.96*se0[0]:.3f}, {beta0[0]+1.96*se0[0]:.3f}]")
print(f"tau2={tau2_0:.4f}, I2={I2_0:.1f}%")
print(f"Subcategory moderator: QM({df_QM})={QM:.2f}, p={p_QM:.4f}")
print(f"R2={R2*100:.1f}%")
print(f"Original KW: H=8.71, p=0.1212 (unweighted, no sampling variances)")
